src/category_theory/limits/shapes/kernel_pair.lean

```diff
@@ -0,0 +1,174 @@
+import category_theory.limits.shapes.pullbacks
+import category_theory.limits.shapes.regular_mono
+
+/-!
+# Kernel pairs
+
+This file defines what it means for a parallel pair of morphisms `a,b : R ⟶ X` to be the kernel pair
+for a morphism `f`.
+Some properties of kernel pairs are given, namely allowing one to transfer between
+the kernel pair of `f₁ ≫ f₂` to the kernel pair of `f₁`.
+It is also proved that If `f` is a coequalizer of some pair, and it is a kernel pair for `a,b` then
+it is a coequalizer of `a,b`.
+
+## Implementation
+
+The definition is essentially just a wrapper for `is_limit (pullback_cone.mk _ _ _)`, but the
+constructions given here are useful, yet awkward to present in that language, so a basic API
+is developed here.
+
+## TODO
+
+- Internal equivalence relations (or congruences) and the fact that every kernel pair induces one,
+  and the converse in an effective regular category (WIP by b-mehta).
+
+-/
+
+universes v u u₂
+
+namespace category_theory
+
+open category_theory category_theory.category category_theory.limits
+
+variables {C : Type u} [category.{v} C]
+
+variables {R X Y Z : C} (f : X ⟶ Y) (a b : R ⟶ X)
+
+/--
+`is_kernel_pair f a b` expresses that `a,b` is a kernel pair for `f`, i.e. `a ≫ f = b ≫ f`
+and the square
+  R → X
+  ↓   ↓
+  X → Y
+is a pullback square.
+This is essentially just a convenience wrapper over `is_limit (pullback_cone.mk _ _ _)`.
+-/
+structure is_kernel_pair :=
+(comm : a ≫ f = b ≫ f)
+(is_limit : is_limit (pullback_cone.mk _ _ comm))
+
+attribute [reassoc] is_kernel_pair.comm
+
+namespace is_kernel_pair
+
+/-- The data expressing that `a,b` is a kernel pair is subsingleton. -/
+instance : subsingleton (is_kernel_pair f a b) :=
+⟨λ P Q, begin
+  cases P,
+  cases Q,
+  congr,
+end⟩
+
+/-- If `f` is a monomorphism, then `𝟙 _, 𝟙 _`  is a kernel pair for `f`. -/
+def id_of_mono [mono f] : is_kernel_pair f (𝟙 _) (𝟙 _) :=
+{ comm := rfl,
+  is_limit :=
+  pullback_cone.is_limit_aux' _ $ λ s,
+  begin
+    refine ⟨s.snd, _, comp_id _, λ m m₁ m₂, _⟩,
+    { rw [← cancel_mono f, s.condition, pullback_cone.mk_fst, cancel_mono f],
+      apply comp_id },
+    rw [← m₂],
+    apply (comp_id _).symm,
+  end }
+
+instance [mono f] : inhabited (is_kernel_pair f (𝟙 _) (𝟙 _)) := ⟨id_of_mono f⟩
+
+variables {f a b}
+
+/--
+Given a pair of morphisms `p,q` to `X` which factor through `f`, they factor through any kernel
+pair of `f`.
+-/
+def lift' {S : C} (k : is_kernel_pair f a b) (p q : S ⟶ X) (w : p ≫ f = q ≫ f) :
+  { t : S ⟶ R // t ≫ a = p ∧ t ≫ b = q } :=
+pullback_cone.is_limit.lift' k.is_limit _ _ w
+
+/--
+If `(a,b)` is a kernel pair for `f₁ ≫ f₂` and `a ≫ f₁ = b ≫ f₁`, then `(a,b)` is a kernel pair for
+just `f₁`.
+That is, to show that `(a,b)` is a kernel pair for `f₁` it suffices to only show the square
+commutes, rather than to additionally show it's a pullback.
+-/
+def sub {f₁ : X ⟶ Y} {f₂ : Y ⟶ Z} (comm : a ≫ f₁ = b ≫ f₁) (big_k : is_kernel_pair (f₁ ≫ f₂) a b) :
```

Comment:
I'd appreciate a name suggestion for this (as well as `sub'` and `sub''` below)!

Reply:
`cancel_right`?